ignite/metrics/nlp/bleu.py

```diff
@@ -0,0 +1,184 @@
+import math
+from collections import Counter
+from typing import Any, Callable, Optional, Sequence, Tuple, Union
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+from ignite.metrics.nlp.utils import modified_precision
+
+__all__ = ["Bleu"]
+
+
+def _closest_ref_length(references: Sequence[Sequence[Any]], hyp_len: int) -> int:
+    ref_lens = (len(reference) for reference in references)
+    closest_ref_len = min(ref_lens, key=lambda ref_len: (abs(ref_len - hyp_len), ref_len))
+    return closest_ref_len
+
+
+class _Smoother:
+    """
+    Smoothing helper
+    http://acl2014.org/acl2014/W14-33/pdf/W14-3346.pdf
+    """
+
+    def __init__(self, method: str):
+        valid = ["no_smooth", "smooth1", "nltk_smooth2", "smooth2"]
+        if method not in valid:
+            raise ValueError(f"Smooth is not valid (expected: {valid}, got: {method})")
+        self.smooth = method
+
+    def __call__(self, numerators: Counter, denominators: Counter) -> Sequence[float]:
+        method = getattr(self, self.smooth)
+        return method(numerators, denominators)
+
+    @staticmethod
+    def smooth1(numerators: Counter, denominators: Counter) -> Sequence[float]:
+        epsilon = 0.1
+        denominators = [max(1, d) for d in denominators.values()]
+        return [n / d if n != 0 else epsilon / d for n, d in zip(numerators.values(), denominators)]
+
+    @staticmethod
+    def nltk_smooth2(numerators: Counter, denominators: Counter) -> Sequence[float]:
+        denominators = [max(1, d) for d in denominators.values()]
+        return [(n + 1) / (d + 1) for n, d in zip(numerators.values(), denominators)]
+
+    @staticmethod
+    def smooth2(numerators: Counter, denominators: Counter) -> Sequence[float]:
```

Comment:
This is not giving different results than nltk_smooth2 for me. Tested it with 3-4 values. Am I missing something?
```
from collections import Counter


def nltk_smooth2(numerators: Counter, denominators: Counter):
    denominators = [max(1, d) for d in denominators.values()]
    return [(n + 1) / (d + 1) for n, d in zip(numerators.values(), denominators)]


def smooth2(numerators: Counter, denominators: Counter):
    return [(n + 1) / (d + 1) for n, d in zip(numerators.values(), denominators.values())]


a = [1, 1, 2, 2, 2, 3, 3, 3, 4, 5]
b = [1, 1, 1, 2, 2, 3, 3, 4, 5, 5]
num = Counter(a)
den = Counter(b)
print(num, den)
for n, d in zip(num.values(), den):
    print(n, d)
print("here")
for n, d in zip(num.values(), den.values()):
    print(n, d)
print(nltk_smooth2(num, den))
print(smooth2(num, den))
```

Reply:
It is similar because you have ngram matching counts.

By default, nltk do for the denominator 

https://github.com/nltk/nltk/blob/1f50455ef3d754996e983699352c4d7bfa8a07cb/nltk/translate/bleu_score.py#L330

So for `smooth2`, if denominator (and denominator) was zero, the precision is : `(0 + 1) / (1 + 1)` rather than `(0 + 0) / (0 + 0)`.

For details https://www.aclweb.org/anthology/W14-3346.pdf

Reply:
No, I mean the nltk_smooth2 and smooth2 give the same result so why do we have both the functions. Can you give me a test case where the result for nltk_smooth2 and smooth2 are different?

Reply:
I also wonder what would happen if counter's keys are different ? Can we encounter such situation : 
```
num = {1: 1, 2: 2}
den = {2: 1, 3: 2}
```

Reply:
@gucifer 

```python
import pytest
import warnings
from nltk.translate.bleu_score import corpus_bleu, SmoothingFunction

from ignite.metrics.nlp import Bleu


with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    reference = corpus_bleu([[["0"]]], [["0"]], smoothing_function=SmoothingFunction().method2)

m = Bleu(ngram=4, smooth="smooth2")

m.update((["0"], [["0"]]))

assert pytest.approx(reference) != m.compute()

m = Bleu(ngram=4, smooth="nltk_smooth2")

m.update((["0"], [["0"]]))

assert pytest.approx(reference) == m.compute()
```

Reply:
> No, I mean the nltk_smooth2 and smooth2 give the same result so why do we have both the functions. Can you give me a test case where the result for nltk_smooth2 and smooth2 are different?



> I also wonder what would happen if counter's keys are different ? Can we encounter such situation :
> 
> ```
> num = {1: 1, 2: 2}
> den = {2: 1, 3: 2}
> ```

It can't happen with the algorithm because operator `&` is used.